### PR TITLE
[BugFix] Fix query partition compensate predicates for mv rewrite 

### PR DESCRIPTION
**Pull request by LiShuMing**
### Main
1. When MV has enough partitions for the query, no need to compensate anymore for both mv and the query's plan.
A query can be rewritten just by the original SQL meaning.
2. If not, we compensate the query and mv's partition ranges by the selected partition ids, so can be used for the `UnionAll` rewrite.

### Changes

1. Add `isNeedCompensatePartitionPredicate ` function to check whether to compensate query or mv partition predicates anymore when MV has enough partitions for the query.
2. Ignore partition predicates that are pushed down or redundant which may be noisy for rewriting.
3. Fix some bugs and add more tests 

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 59 / 63 (93.65%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|22|25|88.00%|[387, 1023, 1027]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|33|34|97.06%|[680]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 63 / 69 (91.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|4|6|66.67%|[116, 123]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|26|29|89.66%|[387, 1024, 1028]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|33|34|97.06%|[680]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 122 / 132 (92.42%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|69|78|88.46%|[413, 932, 938, 945, 960, 980, 986, 998, 1139]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|33|34|97.06%|[680]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewriteContext.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|7|7|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 124 / 135 (91.85%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|69|78|88.46%|[413, 941, 947, 954, 969, 989, 995, 1007, 1148]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|8|9|88.89%|[109]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|34|35|97.14%|[683]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewriteContext.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 1]
```diff
@@ -249,16 +252,22 @@ boolean checkJoinOnPredicate(LogicalJoinOperator mvJoinOp, LogicalJoinOperator q
...
-        final ScalarOperator queryJoinOnPredicate = queryJoinOp.getOriginalOnPredicate();
+        // Checks whether the join-on predicate of a query is equivalent to the join-on predicate
+        // of a materialized view (MV).
+        final ScalarOperator mvJoinOnPredicate = mvJoinOp.getOnPredicate();
+        final ScalarOperator queryJoinOnPredicate = queryJoinOp.getOnPredicate();
         final ScalarOperator normQueryJoinOnPredicate = MvUtils.canonizePredicateForRewrite(queryJoinOnPredicate);
         final ScalarOperator normMVJoinOnPredicate = MvUtils.canonizePredicateForRewrite(mvJoinOnPredicate);
         final Set<ScalarOperator> queryJoinOnPredicates = Sets.newHashSet(Utils.extractConjuncts(normQueryJoinOnPredicate));
         final Set<ScalarOperator> diffPredicates = Sets.newHashSet(Utils.extractConjuncts(normMVJoinOnPredicate));
+        if (!checkJoinOnPredicateFromOnPredicates(queryJoinOnPredicates, diffPredicates)) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:263]: this name checkJoinOnPredicateFromOnPredicates is ambiguous

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java [hunk 2]
```diff
@@ -899,15 +916,111 @@ public static ScalarOperator compensatePartitionPredicate(OptExpression plan,
...
+     * Determine whether to compensate extra partition predicates,
+     * - if it needs compensate, use `selectedPartitionIds` to compensate complete partition ranges
+     *  with lower and upper bound.
+     * - if not compensate, use original pruned partition predicates as the compensated partition
+     *  predicates.
+     * @param plan : query opt expression
+     * @param mvContext : materialized view context
+     * @return
+     */
+    public static boolean isNeedCompensatePartitionPredicate(OptExpression plan,
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java:928]: add some comment to describe the judgement logic

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:263]: fixed

**Review by ABingHuang — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**